net/proxy_resolution/proxy_config_service_tor.cc

```diff
@@ -0,0 +1,194 @@
+/* Copyright (c) 2019 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "brave/net/proxy_resolution/proxy_config_service_tor.h"
+
+#include <algorithm>
+#include <memory>
+#include <utility>
+#include <vector>
+
+#include "base/no_destructor.h"
+#include "base/strings/string_number_conversions.h"
+#include "base/strings/string_util.h"
+#include "base/time/time.h"
+#include "base/values.h"
+#include "crypto/random.h"
+#include "net/base/registry_controlled_domains/registry_controlled_domain.h"
+#include "net/proxy_resolution/proxy_config_with_annotation.h"
+#include "net/proxy_resolution/proxy_resolution_service.h"
+#include "url/origin.h"
+
+namespace net {
+
+namespace {
+
+static base::NoDestructor<std::map<
+    ProxyResolutionService*, void*>> proxy_profile_map_;
+
+static base::NoDestructor<std::map<
+    void*, ProxyConfigServiceTor::TorProxyMap>> profile_tor_map_;
+}  // namespace
+
+const int kTorPasswordLength = 16;
+// Default tor circuit life time is 10 minutes
+constexpr base::TimeDelta kTenMins = base::TimeDelta::FromMinutes(10);
+
+// static
+ProxyConfigServiceTor::TorProxyMap*
+ProxyConfigServiceTor::GetTorProxyMap(ProxyResolutionService* service) {
+  void* profile = (*proxy_profile_map_)[service];
+  if (profile) {
+    return &((*profile_tor_map_)[profile]);
```

Comment:
minor, but for this code I would prefer
(a) avoid getting address of map element, because it doesn't work well if the container type would be changed
(b) Add some DCHECKs that validate invariants - checking that the element we trying to get exists (now it will be silently inserted), etc
(c) switch `map` to `flat_map`
(d) write some boilerplate for `NoDestructor` to be CC-conformant (as `no_destructor.h` suggests)